fast_lio_sam_qn/include/utilities.hpp

```diff
@@ -0,0 +1,170 @@
+  voxelgrid.setInputCloud(pcd_in_ptr);
+  voxelgrid.filter(*pcd_out);
+  return pcd_out;
+}
+
+inline pcl::PointCloud<PointType>::Ptr voxelizePcd(const pcl::PointCloud<PointType>::Ptr &pcd_in, const float voxel_res)
+{
+  static pcl::VoxelGrid<PointType> voxelgrid;
+  voxelgrid.setLeafSize(voxel_res, voxel_res, voxel_res);
+  pcl::PointCloud<PointType>::Ptr pcd_out(new pcl::PointCloud<PointType>);
+  pcd_out->reserve(pcd_in->size());
+  voxelgrid.setInputCloud(pcd_in);
+  voxelgrid.filter(*pcd_out);
+  return pcd_out;
+}
+
+//////////////////////////////////////////////////////////////////////
+///// conversions
+inline gtsam::Pose3 poseEigToGtsamPose(const Eigen::Matrix4d &pose_eig_in)
+{
+	double r, p, y;
+	tf::Matrix3x3 mat;
+	tf::matrixEigenToTF(pose_eig_in.block<3, 3>(0, 0), mat);
+	mat.getRPY(r, p, y);
+	return gtsam::Pose3(gtsam::Rot3::RzRyRx(r, p, y), gtsam::Point3(pose_eig_in(0, 3), pose_eig_in(1, 3), pose_eig_in(2, 3)));
+}
+
+inline Eigen::Matrix4d gtsamPoseToPoseEig(const gtsam::Pose3 &gtsam_pose_in)
+{
+	Eigen::Matrix4d pose_eig_out = Eigen::Matrix4d::Identity();
+	tf::Quaternion quat = tf::createQuaternionFromRPY(gtsam_pose_in.rotation().roll(), gtsam_pose_in.rotation().pitch(), gtsam_pose_in.rotation().yaw());
+	tf::Matrix3x3 mat(quat);
+	Eigen::Matrix3d tmp_rot_mat;
+	tf::matrixTFToEigen(mat, tmp_rot_mat);
+	pose_eig_out.block<3, 3>(0, 0) = tmp_rot_mat;
+	pose_eig_out(0, 3) = gtsam_pose_in.translation().x();
+	pose_eig_out(1, 3) = gtsam_pose_in.translation().y();
+	pose_eig_out(2, 3) = gtsam_pose_in.translation().z();
+	return pose_eig_out;
+}
+
+inline geometry_msgs::PoseStamped poseEigToPoseStamped(const Eigen::Matrix4d &pose_eig_in, std::string frame_id="map")
+{
+	double r, p, y;
+	tf::Matrix3x3 mat;
+	tf::matrixEigenToTF(pose_eig_in.block<3, 3>(0, 0), mat);
+	mat.getRPY(r, p, y);
+	tf::Quaternion quat = tf::createQuaternionFromRPY(r, p, y);
+	geometry_msgs::PoseStamped pose;
+	pose.header.frame_id = frame_id;
+	pose.pose.position.x = pose_eig_in(0, 3);
+	pose.pose.position.y = pose_eig_in(1, 3);
+	pose.pose.position.z = pose_eig_in(2, 3);
+	pose.pose.orientation.w = quat.getW();
+	pose.pose.orientation.x = quat.getX();
+	pose.pose.orientation.y = quat.getY();
+	pose.pose.orientation.z = quat.getZ();
+	return pose;
+}
+
+inline tf::Transform poseEigToROSTf(const Eigen::Matrix4d &pose)
+{
+  Eigen::Quaterniond quat(pose.block<3,3>(0, 0));
+	tf::Transform transform;
+  transform.setOrigin(tf::Vector3(pose(0, 3), pose(1, 3), pose(2, 3)));
+  transform.setRotation(tf::Quaternion(quat.x(), quat.y(), quat.z(), quat.w()));
+	return transform;
+}
+
+inline tf::Transform poseStampedToROSTf(const geometry_msgs::PoseStamped &pose)
+{
+	tf::Transform transform;
+  transform.setOrigin(tf::Vector3(pose.pose.position.x, pose.pose.position.y, pose.pose.position.z));
+  transform.setRotation(tf::Quaternion(pose.pose.orientation.x, pose.pose.orientation.y, pose.pose.orientation.z, pose.pose.orientation.w));
+	return transform;
+}
```

Comment:
![image](https://github.com/user-attachments/assets/123506e5-aab4-4b70-a329-2d301b65c182)

I think spaces are due to the conflict between my editor and your IDE. It could be easily solved if we set .clang-format file.